import sys
import os
import grpc
from google.protobuf import empty_pb2

# Garante que o Python ache os arquivos .py gerados a partir dos .proto
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'proto')))

import voting_pb2
import voting_pb2_grpc


def submit_vote(stub):
    voter_id = input("üÜî Digite seu ID de eleitor: ").strip()
    print("üë§ Escolha um candidato:")
    candidates = {
        1: "Alice",
        2: "Bruno",
        3: "Carlos",
        4: "Daniela"
    }
    for cid, name in candidates.items():
        print(f"{cid}. {name}")

    try:
        choice = int(input("Digite o n√∫mero do candidato: "))
        if choice not in candidates:
            print("‚ùå Op√ß√£o inv√°lida!")
            return
        candidate_id = candidates[choice]  # string conforme o seu proto
    except ValueError:
        print("‚ùå Entrada inv√°lida!")
        return

    try:
        req = voting_pb2.VoteRequest(voter_id=voter_id, candidate_id=candidate_id)
        resp = stub.SubmitVote(req)
        print(f"‚úÖ {resp.message}")
    except grpc.RpcError as e:
        print(f"‚ùå Erro ao enviar voto: {e.details()}")


def get_results(stub):
    try:
        resp = stub.QueryResults(empty_pb2.Empty())
        print("\nüìä Resultados da vota√ß√£o:")
        for candidate, votes in resp.results.items():
            print(f"‚û°Ô∏è {candidate}: {votes} voto(s)")
    except grpc.RpcError as e:
        print(f"‚ùå Erro ao buscar resultados: {e.details()}")


def main():
    with grpc.insecure_channel('localhost:50051') as channel:
        stub = voting_pb2_grpc.VotingServiceStub(channel)

        while True:
            print("\n========= MENU =========")
            print("1. Votar")
            print("2. Ver resultados")
            print("3. Sair")
            print("========================")
            opc = input("Escolha uma op√ß√£o: ").strip()

            if opc == '1':
                submit_vote(stub)
            elif opc == '2':
                get_results(stub)
            elif opc == '3':
                print("üëã Encerrando cliente...")
                break
            else:
                print("‚ùå Op√ß√£o inv√°lida! Tente novamente.")


if __name__ == '__main__':
    main()
